File: src/external_downloader.py
"""Download NadeoImporter and Blendermania_Dotnet to the user's tools dir.

NadeoImporter comes from Nadeo's official CDN (free, no auth, stable URL).
Blendermania_Dotnet comes from the blendermania-assets GitHub release
(the addon team distributes it on their own; we piggyback because we don't
build the .NET helper ourselves).
"""
from __future__ import annotations

import urllib.request
import zipfile
from dataclasses import dataclass
from pathlib import Path

# Latest TM2020-compatible NadeoImporter version (per
# vendor/blendermania-addon/properties/Functions.py:78)
NADEO_IMPORTER_VERSION = "2022_07_12"

# Per vendor/blendermania-addon/utils/Constants.py:206
BLENDERMANIA_DOTNET_VERSION = "v1.0.0"

# Official Nadeo CDN — bypasses any third-party mirror
NADEO_IMPORTER_URL = (
    f"https://nadeo-download.cdn.ubi.com/trackmania/NadeoImporter_{NADEO_IMPORTER_VERSION}.zip"
)

# blendermania-assets mirror — the addon team's own release of their .NET helper
_GITHUB_ASSETS = "https://github.com/skyslide22/blendermania-assets/releases/download"
BLENDERMANIA_DOTNET_URL = (
    f"{_GITHUB_ASSETS}/Blendermania_Dotnet_{BLENDERMANIA_DOTNET_VERSION}/"
    f"Blendermania_Dotnet_{BLENDERMANIA_DOTNET_VERSION}.zip"
)


@dataclass
class DownloadResult:
    name: str
    url: str
    dst_dir: Path
    extracted_files: list[Path]


def _download_to(url: str, dst: Path, progress=None) -> None:
    """Download `url` to `dst`. `progress(bytes_so_far, total_bytes)` callback optional.

    Validates the response is a zip (or at least not HTML) before writing —
    avoids silently saving a 404/error page as a corrupt "zip" the user only
    notices when extraction fails.
    """
    req = urllib.request.Request(url, headers={"User-Agent": "forzamania/0.1"})
    with urllib.request.urlopen(req, timeout=120) as resp:
        ctype = (resp.headers.get("Content-Type") or "").lower()
        if "html" in ctype or "text" in ctype:
            raise RuntimeError(
                f"download URL returned {ctype} (expected zip): {url} — "
                f"the server probably gave a 404 or error page instead of the asset"
            )

        total = int(resp.headers.get("Content-Length", "0") or 0)
        dst.parent.mkdir(parents=True, exist_ok=True)
        bytes_done = 0
        with open(dst, "wb") as f:
            while True:
                chunk = resp.read(1 << 16)
                if not chunk:
                    break
                f.write(chunk)
                bytes_done += len(chunk)
                if progress is not None:
                    progress(bytes_done, total)

    # Magic-byte check: PKZIP files start with "PK\x03\x04" (or PK\x05\x06 for
    # an empty zip). Catches the case where Content-Type lied.
    with open(dst, "rb") as f:
        head = f.read(4)
    if not head.startswith(b"PK"):
        raise RuntimeError(
            f"downloaded file from {url} is not a zip "
            f"(first bytes: {head!r}) — server probably returned an error page"
        )


def _unzip(zip_path: Path, dst_dir: Path) -> list[Path]:
    """Extract `zip_path` into `dst_dir`. Standard PKZIP — no Forza method-21
    weirdness here. Returns the list of files written."""
    dst_dir.mkdir(parents=True, exist_ok=True)
    written: list[Path] = []
    with zipfile.ZipFile(zip_path) as zf:
        for info in zf.infolist():
            if info.is_dir():
                continue
            out = dst_dir / info.filename
            out.parent.mkdir(parents=True, exist_ok=True)
            with zf.open(info) as src, open(out, "wb") as dst:
                dst.write(src.read())
            written.append(out)
    return written


def download_nadeo_importer(
    tools_dir: Path,
    progress=None,
) -> DownloadResult:
    """Download NadeoImporter zip and extract into ``tools_dir``.

    `tools_dir` should typically be ``<TM install>/forzamania-tools/``.
    The blendermania-addon also accepts ``<TM install>/`` directly (canonical
    NadeoImporter location); we use a subfolder so we don't pollute the
    install with our copy.
    """
    tools_dir = Path(tools_dir)
    zip_path = tools_dir / f"NadeoImporter_{NADEO_IMPORTER_VERSION}.zip"
    _download_to(NADEO_IMPORTER_URL, zip_path, progress=progress)
    written = _unzip(zip_path, tools_dir)
    try:
        zip_path.unlink()
    except OSError:
        pass
    return DownloadResult(
        name="NadeoImporter",
        url=NADEO_IMPORTER_URL,
        dst_dir=tools_dir,
        extracted_files=written,
    )


def download_blendermania_dotnet(
    tools_dir: Path,
    progress=None,
) -> DownloadResult:
    """Download Blendermania_Dotnet zip and extract into ``tools_dir``."""
    tools_dir = Path(tools_dir)
    zip_path = tools_dir / f"Blendermania_Dotnet_{BLENDERMANIA_DOTNET_VERSION}.zip"
    _download_to(BLENDERMANIA_DOTNET_URL, zip_path, progress=progress)
    written = _unzip(zip_path, tools_dir)
    try:
        zip_path.unlink()
    except OSError:
        pass
    return DownloadResult(
        name="Blendermania_Dotnet",
        url=BLENDERMANIA_DOTNET_URL,
        dst_dir=tools_dir,
        extracted_files=written,
    )

File: app/external_downloader.nim
## Download nadeo-freeporter to the user's tools dir.
##
## Nim port of `src/external_downloader.py`. Grabs the latest GitHub release and
## picks the asset matching the platform forzamania RUNS AS (freeporter is a
## child process in the same OS-ABI environment): windows→.exe, else→ELF. The
## native Nim build needs HTTPS, so compile this with `-d:ssl` (the Settings
## "Download" button is the only caller; manually pointing the Importer field at
## an existing binary is the offline fallback).
##
## Standard PKZIP (deflate) extraction is delegated to the system `unzip`/`tar`
## — binzip.nim is FM4 method-21-specific and can't inflate. This is a
## click-time convenience, so a system-tool dependency is acceptable here.

import std/[json, os, osproc, strutils, httpclient]

const
  FREEPORTER_REPO = "ItsNotPaths/tm2020-freeporter"
  FREEPORTER_LATEST_API = "https://api.github.com/repos/" & FREEPORTER_REPO & "/releases/latest"

type
  DownloadResult* = object
    name*:    string
    url*:     string
    dstDir*:  string
    binary*:  string   ## path to the extracted platform binary ("" if not found)

proc freeporterAssetForPlatform*(): (string, string) =
  ## (asset_keyword, binary_filename) for the running platform.
  when defined(windows): ("windows", "nadeo-freeporter.exe")
  else:                  ("linux",   "nadeo-freeporter")

proc newGhClient(): HttpClient =
  result = newHttpClient(timeout = 120_000)
  result.headers = newHttpHeaders({
    "User-Agent": "forzamania/0.1",
    "Accept": "application/vnd.github+json",
  })

proc fetchLatestFreeporterAsset(keyword: string): (string, string) =
  ## Query GitHub for the latest release; return (download_url, tag) for the
  ## .zip asset whose name contains `keyword`. Raises if absent.
  let client = newGhClient()
  defer: client.close()
  let data = parseJson(client.getContent(FREEPORTER_LATEST_API))
  let tag = data{"tag_name"}.getStr("?")
  if data.hasKey("assets"):
    for asset in data["assets"]:
      let name = asset{"name"}.getStr("").toLowerAscii
      if keyword in name and name.endsWith(".zip"):
        return (asset["browser_download_url"].getStr, tag)
  raise newException(IOError,
    "no '" & keyword & "' .zip asset in " & FREEPORTER_REPO & " release " & tag)

proc downloadTo(url, dst: string) =
  ## Download `url` to `dst`, then magic-byte check it's a PKZIP (catches a 404/
  ## error page saved as a bogus "zip").
  createDir(parentDir(dst))
  let client = newGhClient()
  defer: client.close()
  client.downloadFile(url, dst)   # follows GitHub's redirect to the asset host
  var head = newString(2)
  let f = open(dst, fmRead)
  defer: f.close()
  let n = f.readBuffer(addr head[0], 2)
  if n < 2 or head != "PK":
    raise newException(IOError,
      "downloaded file from " & url & " is not a zip (first bytes: " & head &
      ") — the server probably returned an error page")

proc extractZip(zipPath, dstDir: string) =
  ## Extract a standard zip via the system `unzip` (preferred) or `tar`.
  createDir(dstDir)
  var rc = -1
  if findExe("unzip").len > 0:
    rc = execCmd("unzip -o -q " & quoteShell(zipPath) & " -d " & quoteShell(dstDir))
  if rc != 0 and findExe("tar").len > 0:
    rc = execCmd("tar -xf " & quoteShell(zipPath) & " -C " & quoteShell(dstDir))
  if rc != 0:
    raise newException(IOError,
      "failed to extract " & zipPath & " (need `unzip` or `tar` on PATH)")

proc downloadFreeporter*(toolsDir: string): DownloadResult =
  ## Download the latest nadeo-freeporter for this platform into `toolsDir`,
  ## extract it, and restore the ELF's exec bit (zip extraction drops it).
  let (keyword, binaryName) = freeporterAssetForPlatform()
  let (url, tag) = fetchLatestFreeporterAsset(keyword)

  createDir(toolsDir)
  let zipPath = toolsDir / ("nadeo-freeporter-" & keyword & "-" & tag & ".zip")
  downloadTo(url, zipPath)
  extractZip(zipPath, toolsDir)
  try: removeFile(zipPath)
  except OSError: discard

  var binary = toolsDir / binaryName
  if not fileExists(binary): binary = ""
  when not defined(windows):
    if binary.len > 0:
      # zip extraction can drop the exec bit; the runner must be able to spawn it.
      var perms = getFilePermissions(binary)
      perms.incl({fpUserExec, fpGroupExec, fpOthersExec})
      setFilePermissions(binary, perms)

  DownloadResult(name: "nadeo-freeporter", url: url, dstDir: toolsDir, binary: binary)
